From 9b4dc58a63c7c561267e846478ef520f4d292c91 Mon Sep 17 00:00:00 2001
From: Cemu <cemu@example.invalid>
Date: Thu, 3 Sep 2026 18:15:00 -0600
Subject: [PATCH 9/9] nl80211: translate Tendonin AP crypto selectors

Keep Nintendo's vendor selectors in the RSN IE generated by hostapd, but use
the equivalent standard PSK and CCMP selectors in nl80211 attributes. Linux
validates START_AP crypto attributes against the wiphy capability list, where
normal SoftMAC drivers advertise only the standard selectors.
---
 src/drivers/driver_nl80211.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index bb77b82..f1c6872 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -3297,7 +3297,11 @@ static u32 wpa_cipher_to_cipher_suite(unsigned int cipher)
 	case WPA_CIPHER_GCMP_256:
 		return RSN_CIPHER_SUITE_GCMP_256;
 	case WPA_CIPHER_CCMP:
+#ifdef CONFIG_TENDONIN
+		return RSN_SELECTOR(0x00, 0x0f, 0xac, 4);
+#else
 		return RSN_CIPHER_SUITE_CCMP;
+#endif /* CONFIG_TENDONIN */
 	case WPA_CIPHER_GCMP:
 		return RSN_CIPHER_SUITE_GCMP;
 	case WPA_CIPHER_TKIP:
@@ -3325,7 +3329,11 @@ static int wpa_cipher_to_cipher_suites(unsigned int ciphers, u32 suites[],
 	if (num_suites < max_suites && ciphers & WPA_CIPHER_GCMP_256)
 		suites[num_suites++] = RSN_CIPHER_SUITE_GCMP_256;
 	if (num_suites < max_suites && ciphers & WPA_CIPHER_CCMP)
+#ifdef CONFIG_TENDONIN
+		suites[num_suites++] = RSN_SELECTOR(0x00, 0x0f, 0xac, 4);
+#else
 		suites[num_suites++] = RSN_CIPHER_SUITE_CCMP;
+#endif /* CONFIG_TENDONIN */
 	if (num_suites < max_suites && ciphers & WPA_CIPHER_GCMP)
 		suites[num_suites++] = RSN_CIPHER_SUITE_GCMP;
 	if (num_suites < max_suites && ciphers & WPA_CIPHER_TKIP)
@@ -3351,7 +3359,13 @@ static int wpa_key_mgmt_to_suites(unsigned int key_mgmt_suites, u32 suites[],
 	    (key_mgmt_suites & (WPA_KEY_MGMT_ ## a))) \
 		suites[num_suites++] = (RSN_AUTH_KEY_MGMT_ ## b)
 	__AKM(IEEE8021X, UNSPEC_802_1X);
+#ifdef CONFIG_TENDONIN
+	if (num_suites < max_suites &&
+	    (key_mgmt_suites & WPA_KEY_MGMT_PSK))
+		suites[num_suites++] = RSN_SELECTOR(0x00, 0x0f, 0xac, 2);
+#else
 	__AKM(PSK, PSK_OVER_802_1X);
+#endif /* CONFIG_TENDONIN */
 	__AKM(FT_IEEE8021X, FT_802_1X);
 	__AKM(FT_PSK, FT_PSK);
 	__AKM(IEEE8021X_SHA256, 802_1X_SHA256);
-- 
2.54.0

